### PR TITLE
Create custom dev reload api route

### PR DESCRIPTION
**Pull request by RubenSandwich**
## 🔗 Relevant links

No preview link as this will purposely **only** work in local dev.

## 🗒️ What

Creates a new **App Router API** route /api/refresh that keeps a connection open on clients so that when they are send the reload message they can reload their webpage.

This makes editing documents in Unified Docs a much better experience.

## 🧪 Testing

1. Checkout this repo
2. Set the following envs
    1. `HASHI_ENV="unified-docs-sandbox"`
    2. `UNIFIED_DOCS_API="http://localhost:8080"`
3. [Checkout this PR from unified docs](https://github.com/hashicorp/web-unified-docs/pull/58)
4. Start both dev-portal and unified docs
    1. dev-portal: `npm run start`
    2. unified-docs: `npm run dev`
5. Visit a page
    1. e.g. `http://localhost:3000/terraform/cli`)
6. Open the DevTools and make sure the Reload Client has connected
    1. e.g. `Reload Client ad581121-b944-495f-aed2-d65fc515b074 Connected`
7. Edit that page in Unified Docs
    1. example page is contained in: `content/terraform/v1.8.x/docs/cli/index.mdx`
8. The page should reload

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
 			"@build-libs/*": ["../build-libs/*"],
 			"@components/*": ["components/*"]
 		},
-		"types": ["vitest/globals", "@testing-library/jest-dom"]
+		"types": ["vitest/globals", "@testing-library/jest-dom"],
+		"strictNullChecks": false
```
Comment (RubenSandwich) [inline tsconfig.json:26]: This was required to get vercel to build for me, or else it was injecting `"strictNullChecks": true` 😢

File: src/app/api/refresh/route.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import { NextRequest } from 'next/server'
```
Comment (heatlikeheatwave) [inline src/app/api/refresh/route.ts:1]:
```suggestion
/**
 * Copyright (c) HashiCorp, Inc.
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 * SPDX-License-Identifier: MPL-2.0
 */
 */
 
import { NextRequest } from 'next/server'
```

File: src/app/api/refresh/route.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+type Client = {
+	id: string
+	res: any
```
Comment (heatlikeheatwave) [inline src/app/api/refresh/route.ts:5]: Can it be something other than `any`?

File: src/app/api/refresh/route.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+const clients: Client[] = []
+
+function addClient(id: string, res: any): void {
+	clients.push({ id, res })
+}
+
+function removeClient(id: string): void {
+	const index = clients.findIndex((client) => client.id === id)
```
Comment (heatlikeheatwave) [inline src/app/api/refresh/route.ts:15]: should we add a lil check for clients have elements here?

Comment (RubenSandwich) [reply src/app/api/refresh/route.ts:15]: Huh, what do you mean?

File: src/app/api/refresh/route.ts (hunk 3 above)
Comment (RubenSandwich) [reply src/app/api/refresh/route.ts:5]: It's technically this: https://github.com/hashicorp/dev-portal/pull/2615/files#diff-5f0aec67bf02b90279a7a6c4009675ff4574bf755ced6fc905252a7c495e5c70R54.

But I got lazy, especially with just a dev tool.

File: src/app/api/refresh/route.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+function sendMessageToAllClients(obj: any): void {
+	const jsonData = `data: ${JSON.stringify(obj)}\n\n`
+	for (const client of clients) {
+		client.res.write(jsonData)
+	}
+}
+
+export async function POST() {
+	if (process.env.NODE_ENV !== 'development') {
```
Comment (heatlikeheatwave) [inline src/app/api/refresh/route.ts:34]:
```suggestion
	if (process.env.VERCEL_ENV !== 'development') {
```

File: src/app/api/refresh/route.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		return new Response('Not Found', { status: 404 })
+	}
+
+	sendMessageToAllClients({ reload: true })
+	return new Response('Reload sent to all clients')
+}
+
+export async function GET(req: NextRequest) {
+	if (process.env.NODE_ENV !== 'development') {
```
Comment (heatlikeheatwave) [inline src/app/api/refresh/route.ts:43]:
```suggestion
	if (process.env.VERCEL_ENV !== 'development') {
```

File: src/app/api/refresh/route.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	})
+
+	req.signal.addEventListener('abort', () => {
+		console.log(`Reload Client ${clientId} Disconnected`)
+		removeClient(clientId)
+		writer.close()
+	})
+
+	// Send to start the connection
+	writer.write(encoder.encode('data: \n\n'))
```
Comment (RubenSandwich) [inline src/app/api/refresh/route.ts:79]: Believe it or not the two endlines are required for it to work. 🤷🏻

File: src/layouts/base-layout/index.tsx [hunk 8]
```diff
@@ -58,6 +58,34 @@ const BaseLayout = ({
 	useScrollPercentageAnalytics()
 	const [showSkipLink, setShowSkipLink] = useState(false)
 
+	if (process.env.NODE_ENV === 'development') {
```
Comment (heatlikeheatwave) [inline src/layouts/base-layout/index.tsx]:
```suggestion
	if (process.env.VERCEL_ENV === 'development') {
```

File: src/layouts/base-layout/index.tsx [hunk 9]
```diff
@@ -58,6 +58,34 @@ const BaseLayout = ({
...
+			}
+
+			eventSource.onmessage = (event) => {
+				const data = event.data && JSON.parse(event?.data)
+
+				if (data.reload) {
+					console.log('Reload Client Reloading Page')
+					window.location.reload()
+				}
+			}
```
Comment (heatlikeheatwave) [inline src/layouts/base-layout/index.tsx:77]: The same, but in a try/catch block.
```suggestion
				try {
				  const data = event.data && JSON.parse(event?.data)
  
				  if (data.reload) {
					  console.log('Reload Client Reloading Page')
					  window.location.reload()
				  }
				} catch (err) {
					console.error('Error parsing message data:', err);
				}
```

File: src/layouts/base-layout/index.tsx [hunk 10]
```diff
@@ -58,6 +58,34 @@ const BaseLayout = ({
 	useScrollPercentageAnalytics()
 	const [showSkipLink, setShowSkipLink] = useState(false)
 
+	if (process.env.VERCEL_ENV === 'development') {
```
Comment (heatlikeheatwave) [inline src/layouts/base-layout/index.tsx:61]:
```suggestion
	if (process.env.VERCEL_ENV === 'development' && process.env.HASHI_ENV === 'unified-docs-sandbox') {
```

**Review by heatlikeheatwave — APPROVED** [review]

File: src/app/api/refresh/route.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+	for (const client of clients) {
+		client.res.write(jsonData)
+	}
+}
+
+export async function POST() {
+	if (
+		process.env.VERCEL_ENV !== 'development' ||
+		process.env.HASHI_ENV !== 'unified-docs-sandbox'
+	) {
```
Comment (heatlikeheatwave) [inline src/app/api/refresh/route.ts:46]: nice

**Pull request merged**